packages/adblocker/src/engine/bucket/preprocessor.ts

```diff
@@ -0,0 +1,149 @@
+import { StaticDataView } from '../../data-view';
+import IFilter from '../../filters/interface';
+import Preprocessor, { Env } from '../../preprocessor';
+
+export default class PreprocessorBucket {
+  public static deserialize(view: StaticDataView): PreprocessorBucket {
+    const preprocessors: Preprocessor[] = [];
+
+    for (let i = 0, l = view.getUint32(); i < l; i++) {
+      preprocessors.push(Preprocessor.deserialize(view));
+    }
+
+    return new this({
+      preprocessors,
+    });
+  }
+
+  private readonly preprocessors: Preprocessor[];
+  private readonly excluded: Set<number>;
+
+  constructor({
+    env = new Env(),
+    preprocessors = [],
+  }: {
+    env?: Env;
+    preprocessors?: Preprocessor[];
+  }) {
+    this.excluded = new Set();
+    this.preprocessors = preprocessors;
+
+    if (preprocessors.length) {
+      this.update({ added: preprocessors, env });
+    }
+  }
+
+  private build(env: Env) {
+    // Update excluded filter ids based on bindings
+    this.excluded.clear();
+
+    const allowed = new Set<number>();
+
+    for (let i = 0; i < this.preprocessors.length; ) {
+      if (this.preprocessors[i].filterIDs.size === 0) {
+        this.preprocessors.splice(i, 1);
+
+        continue;
+      }
+
+      if (this.preprocessors[i].evaluate(env)) {
+        for (const filterID of this.preprocessors[i].filterIDs) {
+          allowed.add(filterID);
+        }
+      } else {
+        for (const filterID of this.preprocessors[i].filterIDs) {
+          if (!allowed.has(filterID)) {
+            this.excluded.add(filterID);
+          }
+        }
+      }
+
+      i++;
+    }
+  }
+
+  public isFilterExcluded(filter: IFilter) {
+    return this.excluded.has(filter.getId());
+  }
+
+  public updateEnv(env: Env) {
+    this.build(env);
+  }
+
+  public update({
```

Comment:
Was this method already updated to account for the new format of `generateDiffs`?

Reply:
I wasn't aware of diffing functions when I code this method. I just followed other update functions. We may change this to accept the output type from "generateDiff".

Reply:
I found that this function should not follow the format of `generateDiff` because it will force us to parse the filters again. 

```ts
interface IBaseDiff {
  added: string[];
  removed: string[];
}

export interface IPreprocessorDiff {
  [key: string]: IBaseDiff;
}
```